tardis/plasma/properties/nlte_rate_equation_solver.py

```diff
@@ -558,3 +584,114 @@ def prepare_first_guess(
             index += atomic_number + 1
         first_guess[-1] = electron_density
         return first_guess
+
+    def population_objective_function(
+        self,
+        populations,
+        atomic_numbers,
+        phi,
+        solution_vector,
+        rate_matrix_index,
+        total_photo_ion_coefficients,
+        total_rad_recomb_coefficients,
+        total_coll_ion_coefficients,
+        total_coll_recomb_coefficients,
+    ):
+        """Main set of equations for the NLTE ionization solver. A*x - B = 0, where x is the populations,
+        A is the matrix of rates and B is the solution vector.
+
+        Parameters
+        ----------
+        populations : numpy.array
+            Current values of ion number densities and electron density.
+        atomic_numbers : numpy.array
+            All atomic numbers present in the plasma.
+        phi :DataFrame
+            Saha Factors of the current shell.
+        solution_vector : numpy.array
+            Solution vector for the set of equations.
+        rate_matrix_index : MultiIndex
+            (atomic_number, ion_number, treatment type)
+            If ion is treated in LTE or nebular ionization, 3rd index is "lte_ion",
+            if treated in NLTE ionization, 3rd index is "nlte_ion".
+        total_photo_ion_coefficients : DataFrame
+            Photo ion. coefficients for current atomic number
+        total_rad_recomb_coefficients : DataFrame
+            Rad. recomb. coefficients for current atomic number
+        total_coll_ion_coefficients : DataFrame
+            Coll. ion. coefficients for current atomic number
+        total_coll_recomb_coefficients : DataFrame
+            Coll. recomb. coefficients for current atomic number
+        Returns
+        -------
+        _type_
+            _description_
+        """
+        electron_density = populations[-1]
+        rate_matrix = self.calculate_rate_matrix(
+            atomic_numbers,
+            phi,
+            electron_density,
+            rate_matrix_index,
+            total_photo_ion_coefficients,
+            total_rad_recomb_coefficients,
+            total_coll_ion_coefficients,
+            total_coll_recomb_coefficients,
+        )
+        jacobian_matrix = self.jacobian_matrix(
+            atomic_numbers,
+            populations,
+            rate_matrix,
+            rate_matrix_index,
+            total_rad_recomb_coefficients,
+            total_coll_ion_coefficients,
+            total_coll_recomb_coefficients,
+        )
+        return (
+            np.dot(rate_matrix.values, populations) - solution_vector,
+            jacobian_matrix,
+        )
+
+    def solution_vector_block(self, atomic_number, number_density):
+        """Block of the solution vector corresponding to the current atomic number. (0, 0, ..., number_density). Length is equal to atomic_number+1.
```

Comment:
I would suggest to break this up into a "short summary" and an "extended summary" as described in the [pandas docstring guide](https://pandas.pydata.org/docs/development/contributing_docstring.html).